Add more visibility and control around reactions/flags into the Admin USERS page

rt4914 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://developers.forem.com/contributing-guide/forem#create-a-pull-request
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/main/CODE_OF_CONDUCT.md
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.

     NOTE: Pull Requests from forked repositories will need to be reviewed by
     a Forem Team member before any CI builds will run. Once your PR is approved
     with a `/ci` reply to the PR, it will be allowed to run subsequent builds without
     manual approval.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

## Related Tickets & Documents

<!--
For pull requests that relate or close an issue, please include them
below.  We like to follow [Github's guidance on linking issues to pull requests](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

For example having the text: "closes #1234" would connect the current pull
request to issue 1234.  And when we merge the pull request, Github will
automatically close the issue.
-->

- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

_Please replace this line with instructions on how to test your changes, a note
on the devices and browsers this has been tested on, as well as any relevant
images for UI changes._

### UI accessibility checklist
_If your PR includes UI changes, please utilize this checklist:_
- [ ] Semantic HTML implemented?
- [ ] Keyboard operability supported?
- [ ] Checked with [axe DevTools](https://www.deque.com/axe/) and addressed `Critical` and `Serious` issues?
- [ ] Color contrast tested?

_For more info, check out the
[Forem Accessibility Docs](https://developers.forem.com/frontend/accessibility)._

## Added/updated tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

anujbhavsar96 commented:
I tested this in Uffizzi. Working as expected 🚀 @rt4914 

Just two small design suggestions,

1. Can we increase the gap between the title and flag+score summary components on the right for the **Desktop only**? It's just to balance the sentence visually balanced.
2. Can we keep the side and top padding the same for the summer component? 24px top and right.

| Before | After |
| ------- | ------ |
| <img width="892" alt="Screenshot 2023-06-01 at 8 31 58 AM" src="https://github.com/forem/forem/assets/84626054/22569dbc-f12b-41be-a361-2e8d79cbf556"> | <img width="896" alt="Screenshot 2023-06-01 at 8 31 43 AM" src="https://github.com/forem/forem/assets/84626054/0e7c7b78-03bc-4d8a-9b95-71ce4104591d"> |

rt4914 commented:
> I tested this in Uffizzi. Working as expected 🚀 @rt4914
> 
> Just two small design suggestions,
> 
> 1. Can we increase the gap between the title and flag+score summary components on the right for the **Desktop only**? It's just to balance the sentence visually balanced.
> 2. Can we keep the side and top padding the same for the summer component? 24px top and right.

@anujbhavsar96 
I have applied those changes:
<img width="750" alt="Screenshot 2023-06-02 at 1 27 06 PM" src="https://github.com/forem/forem/assets/9396084/eacff576-1754-4964-9051-65ef6f390a06">

rt4914 commented:
@Ridhwana @anujbhavsar96 @dukegreene 
This PR is ready for review. Thanks :-)

mirie commented:
@dukegreene please review this PR. Thanks!

mirie commented:
@rt4914 is this ready for review or is this pending changes from design feedback?

rt4914 commented:
@mirie Yes, this PR is ready for review. There is [one comment](https://github.com/forem/forem/pull/19558#issuecomment-1573659240) from Anuj which needs to be investigated but mostly its not related to this PR.

anujbhavsar96 reviewed (approved):
Design-wise this looks great 🚀

dukegreene reviewed (approved):
Looks good! Was able to flag user #6 with an admin and a trusted user and see the different states of validity (including the initial yellow) on the flags in the admin view.

rt4914 commented:
@anujbhavsar96 Apologies for delay but it was really hard to find issue. So basically what you mentioned earlier about the issue with scoring, it is part of this PR only.
**How do I know that?** Basically I marked user as `Flagged` and `Unflagged` from there profile and checked the score in admin-articles and it was infact turning to `0`. Which is correct and expected.
**So what is the issue?** The issue is that we change the status of flag to `Valid / Invalid / Confirmed`. This is the reason why the scoring is incorrect.
**How to fix it?** Currently I am looking into it. The overall implementation is correct but the score needs to fixed by checking the implementation at other places, mainly to identify what should be the `status` for an `unflagged` user because keeping it as `Invalid` is not a solution.

cc: @Ridhwana  -- you can avoid reviewing this PR for now. I will update you once its ready. And this reply is in context with [this comment](https://github.com/forem/forem/pull/19558#issuecomment-1573659240).

rt4914 commented:
@Ridhwana This PR is ready for review, the issue which Anuj mentioned is not a part of this PR. Also, I have filed the issue for the same. https://github.com/forem/forem/issues/19588

mirie commented:
@Ridhwana if you're able to do the review Friday please note here. Otherwise, if there's another person suitable to nominate to give a review, can you coordinate that? Thanks! It looks like we are so close!!

Ridhwana commented on app/views/admin/articles/_individual_article.html.erb:
```diff
@@ -1,5 +1,5 @@
 <% decorated_article = article.decorate %>
-<%= javascript_packs_with_chunks_tag "admin/articles/vomitReactionItem", defer: true %>
+<%= javascript_packs_with_chunks_tag "admin/shared/flagReactionItemDropdowmButton", defer: true %>
```
typo here and on the name of the file
```suggestion
<%= javascript_packs_with_chunks_tag "admin/shared/flagReactionItemDropdownButton", defer: true %>
```

rt4914 replied:
Done. Thanks.

Ridhwana commented on app/controllers/admin/users_controller.rb:
```diff
@@ -363,7 +363,7 @@ def set_related_reactions
         Reaction.where(reactable_type: "Comment", reactable_id: user_comment_ids, category: "vomit")
           .or(Reaction.where(reactable_type: "Article", reactable_id: user_article_ids, category: "vomit"))
           .or(Reaction.where(reactable_type: "User", reactable_id: @user.id, category: "vomit"))
-          .includes(:reactable)
```
Question: Why do we remove the inclusion of `reactable` here?

rt4914 replied:
Removed this as it was leading to `Avoid eager loading` error.
<img width="553" alt="Screenshot 2023-06-09 at 9 43 50 PM" src="https://github.com/forem/forem/assets/9396084/12c03806-e551-43cb-ab4a-fcc3e335a632">

Ridhwana commented on app/views/admin/users/show/flags/_index.html.erb:
```diff
@@ -1,43 +1,46 @@
-<h2 class="crayons-subtitle-1">Flags received (<%= @related_vomit_reactions.count %>)</h2>
+<%= javascript_packs_with_chunks_tag "admin/shared/flagReactionItemDropdowmButton", defer: true %>
+<% user_vomit_reactions = @related_vomit_reactions.select { |flag| flag.reactable_type == "User" } %>
+<% score = 0 %>
+<% countable_flags = 0 %>
+<% if user_vomit_reactions.present? %>
+  <% user_vomit_reactions.each do |flag| %>
+    <% score += flag.points %>
+      <% if flag.status != 'invalid' %>
+        <% countable_flags += 1 %>
+      <% end %>
+  <% end %>
+<% end %>
```
Is this logic not possible to add to a method in the controller action, or is there a reason that we add it in the template?

rt4914 replied:
Shifted this logic to controller file.

Ridhwana commented on app/views/admin/users/show/flags/_index.html.erb:
```diff
@@ -1,43 +1,46 @@
+  <% if user_vomit_reactions.present? %>
+    <% user_vomit_reactions.each do |vomit_reaction| %>
+      <%= render "admin/shared/flag_reaction_item", vomit_reaction: vomit_reaction, text_section: "users" %>
+      <hr id="js__reaction__div__hr__<%= vomit_reaction.id %>" class="w-100 hr-no-margins">
```
I know that you didn't add this right now but this ID has me so confused  😅

rt4914 replied:
@Ridhwana I do understand this, I just kept it like this considering that it has been used like this in other places too.
Honestly, even I don't know what the exact convention is for naming it, just wanted to keep it consistent with other similar code.

Ridhwana reviewed (commented):
This functionally works well but my main concern is that there is a lot of repetition of html between:

- [app/views/admin/articles/_individual_article.html.erb ](https://github.com/forem/forem/blob/ab5815bfb6452ecccead92ad0b0034597fb984f2/app/views/admin/articles/_individual_article.html.erb )
- [app/views/admin/articles/_article_item.html.erb](https://github.com/forem/forem/blob/bbb9fd2e1e43530a4394441e5625aa9eb88fb0ce/app/views/admin/articles/_article_item.html.erb)

and also within this file and the above two 
- [app/views/admin/users/show/flags/_index.html.erb](https://github.com/forem/forem/blob/ab5815bfb6452ecccead92ad0b0034597fb984f2/app/views/admin/users/show/flags/_index.html.erb)

rt4914 commented:
This PR needs work on 2 things: (1). Update code as per comment https://github.com/forem/forem/pull/19558#discussion_r1224363114
and (2.) the cypress tests are failing on CI but are passing locally, so need to investigate this.

rt4914 commented:
> This functionally works well but my main concern is that there is a lot of repetition of html between:
> 
> * [app/views/admin/articles/_individual_article.html.erb ](https://github.com/forem/forem/blob/ab5815bfb6452ecccead92ad0b0034597fb984f2/app/views/admin/articles/_individual_article.html.erb)
> * [app/views/admin/articles/_article_item.html.erb](https://github.com/forem/forem/blob/bbb9fd2e1e43530a4394441e5625aa9eb88fb0ce/app/views/admin/articles/_article_item.html.erb)
> 
> and also within this file and the above two
> 
> * [app/views/admin/users/show/flags/_index.html.erb](https://github.com/forem/forem/blob/ab5815bfb6452ecccead92ad0b0034597fb984f2/app/views/admin/users/show/flags/_index.html.erb)

@Ridhwana Thanks for the suggestions.
I have removed the `individual_article` file and kept only `article_item` which can be used for both the purposes.
Also, I have extracted the code from `article_item` and `flags/index` into a new file, `/shared/_flag_reaction_table`, this removes the duplicity to a large extent, also keeping the code simple to understand.

This is ready for review now. Thanks.

Ridhwana reviewed (commented):
Thank you for making these improvements @rt4914 ✨ I left a few more non blocking suggestions.

rt4914 commented on app/controllers/admin/articles_controller.rb:
```diff
@@ -30,10 +30,16 @@ def index
 
     def show
       @article = Article.includes(reactions: :user).find(params[:id])
+      @countable_vomits = { @article.id => calculate_flags_for_single_article(@article) }
```
Applied this change.

rt4914 commented on app/controllers/admin/articles_controller.rb:
```diff
@@ -126,5 +132,21 @@ def article_params
+        reaction.reactable_type == "Article"
+      end
+      vomit_article_reactions = privileged_article_reactions.select { |reaction| reaction.category == "vomit" }
+      countable_vomits = 0
```
Changed name to `vomit_count` for better understanding.

rt4914 commented on app/views/admin/articles/_article_item.html.erb:
```diff
@@ -1,203 +1,241 @@
+
+          <span class="crayons-card crayons-card--secondary px-3 py-1 flex gap-2 items-center" title="<%= t("views.moderations.actions.vomit") %>">
+            <%= crayons_icon_tag("twemoji/suspicious", native: true, width: 16, height: 16) %>
+            <span class="fs-s fw-medium lh-base"><%= countable_vomits %></span>
```
Applied this change.
